File: modules/subdivision/examples/mesh_subdivision.cpp
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
#include <lagrange/io/load_mesh.h>
#include <lagrange/io/load_simple_scene.h>
#include <lagrange/io/save_mesh.h>

#include <lagrange/AttributeValueType.h>
#include <lagrange/compute_normal.h>
#include <lagrange/foreach_attribute.h>
#include <lagrange/map_attribute.h>
#include <lagrange/subdivision/compute_sharpness.h>
#include <lagrange/subdivision/mesh_subdivision.h>
#include <lagrange/subdivision/midpoint_subdivision.h>
#include <lagrange/subdivision/sqrt_subdivision.h>
#include <lagrange/topology.h>
#include <lagrange/utils/assert.h>
#include <lagrange/utils/fmt_eigen.h>
#include <lagrange/utils/utils.h>
#include <lagrange/views.h>
#include <lagrange/weld_indexed_attribute.h>

#include <CLI/CLI.hpp>

#include <map>

int main(int argc, char** argv)
{
    struct
    {
        std::string input;
        std::string output = "output.obj";
        std::string scheme = "auto";
        bool output_btn = false;
        std::optional<float> autodetect_normal_threshold_deg;
        int log_level = 1; // debug
    } args;

    lagrange::subdivision::SubdivisionOptions subdivision_options;
    lagrange::subdivision::SharpnessOptions sharpness_options;

    std::map<std::string, lagrange::subdivision::RefinementType> map{
        {"Uniform", lagrange::subdivision::RefinementType::Uniform},
        {"EdgeAdaptive", lagrange::subdivision::RefinementType::EdgeAdaptive}};

    CLI::App app{argv[0]};
    app.option_defaults()->always_capture_default();
    app.add_option("input", args.input, "Input mesh.")->required()->check(CLI::ExistingFile);
    app.add_option("output", args.output, "Output mesh.");
    app.add_option("-s,--scheme", args.scheme, "Subdivision scheme")
        ->check(CLI::IsMember({"auto", "bilinear", "loop", "catmark", "sqrt", "midpoint"}));
    app.add_option(
        "-n,--num-levels",
        subdivision_options.num_levels,
        "Number of subdivision levels");
    app.add_option(
        "-a,--autodetect-normal-threshold",
        args.autodetect_normal_threshold_deg,
        "Normal angle threshold (in degree) for autodetecting sharp edges");
    app.add_flag(
        "--limit",
        subdivision_options.use_limit_surface,
        "Project vertex attributes to the limit surface");
    app.add_option("--refinement", subdivision_options.refinement, "Mesh refinement method")
        ->transform(CLI::CheckedTransformer(map, CLI::ignore_case));
    app.add_option(
        "--edge-length",
        subdivision_options.max_edge_length,
        "Max edge length target for adaptive refinement");
    app.add_flag("--normal", args.output_btn, "Compute limit normal as a vertex attribute");
    app.add_option("-l,--level", args.log_level, "Log level (0 = most verbose, 6 = off).");
    CLI11_PARSE(app, argc, argv);

    args.log_level = std::max(0, std::min(6, args.log_level));
    spdlog::set_level(static_cast<spdlog::level::level_enum>(args.log_level));

    /////////////////////
    // Load input mesh //
    /////////////////////

    lagrange::logger().info("Loading input mesh: {}", args.input);
    lagrange::SurfaceMesh32d mesh;

    bool is_gltf = std::set<std::string>{".gltf", ".glb"}.count(
        lagrange::fs::path(args.input).extension().string());
    if (is_gltf) {
        lagrange::logger().warn(
            "Input mesh is a glTF file. Essential connectivity information is lost when loading "
            "from a glTF asset. We strongly advise using .fbx or .obj as an input file format "
            "rather than glTF.");
    }
    lagrange::io::LoadOptions load_options;
    load_options.stitch_vertices = true;
    mesh = lagrange::io::load_mesh<lagrange::SurfaceMesh32d>(args.input, load_options);
    lagrange::logger().info(
        "Input mesh has {} vertices and {} facets",
        mesh.get_num_vertices(),
        mesh.get_num_facets());

    ///////////////////////
    // Asset preparation //
    ///////////////////////

    // Weld any indexed attribute that we find
    seq_foreach_named_attribute_read(mesh, [&](auto name, auto&& attr) {
        if (mesh.attr_name_is_reserved(name)) {
            return;
        }
        if (attr.get_element_type() == lagrange::AttributeElement::Indexed) {
            lagrange::logger().info("Welding indexed attribute: {}", name);
            lagrange::WeldOptions w_opts;
            w_opts.epsilon_rel = 1e-3;
            w_opts.epsilon_abs = 1e-3;
            lagrange::weld_indexed_attribute(mesh, mesh.get_attribute_id(name), w_opts);
        }
    });
    // Compute sharpness information
    if (args.autodetect_normal_threshold_deg.has_value()) {
        sharpness_options.feature_angle_threshold =
            lagrange::to_radians(args.autodetect_normal_threshold_deg.value());
    }
    auto sharpness_results = lagrange::subdivision::compute_sharpness(mesh, sharpness_options);
    subdivision_options.vertex_sharpness_attr = sharpness_results.vertex_sharpness_attr;
    subdivision_options.edge_sharpness_attr = sharpness_results.edge_sharpness_attr;

    //////////////////////
    // Mesh subdivision //
    //////////////////////

    if ((std::set<std::string>{"auto", "bilinear", "loop", "catmark"}).count(args.scheme)) {
        // Convert subdiv scheme to enum
        if (args.scheme == "loop") {
            subdivision_options.scheme = lagrange::subdivision::SchemeType::Loop;
        } else if (args.scheme == "catmark") {
            subdivision_options.scheme = lagrange::subdivision::SchemeType::CatmullClark;
        } else if (args.scheme == "bilinear") {
            subdivision_options.scheme = lagrange::subdivision::SchemeType::Bilinear;
        }

        if (args.output_btn && sharpness_results.normal_attr.has_value()) {
            // Only output a single set of normals in this example
            mesh.delete_attribute(mesh.get_attribute_name(sharpness_results.normal_attr.value()));
        }

        if (args.output_btn) {
            subdivision_options.output_limit_normals = "normal";
        }
        mesh = lagrange::subdivision::subdivide_mesh(mesh, subdivision_options);
        if (args.output_btn) {
            map_attribute_in_place(mesh, "normal", lagrange::AttributeElement::Indexed);
        }
    } else if (args.scheme == "sqrt") {
        for (unsigned i = 0; i < subdivision_options.num_levels; ++i) {
            mesh = lagrange::subdivision::sqrt_subdivision(mesh);
            if (i + 1 < subdivision_options.num_levels) {
                lagrange::logger().info(
                    "Intermediate mesh has {} vertices and {} facets",
                    mesh.get_num_vertices(),
                    mesh.get_num_facets());
            }
        }
    } else if (args.scheme == "midpoint") {
        for (unsigned i = 0; i < subdivision_options.num_levels; ++i) {
            mesh = lagrange::subdivision::midpoint_subdivision(mesh);
            if (i + 1 < subdivision_options.num_levels) {
                lagrange::logger().info(
                    "Intermediate mesh has {} vertices and {} facets",
                    mesh.get_num_vertices(),
                    mesh.get_num_facets());
            }
        }
    } else {
        throw std::runtime_error("Unsupported argument");
    }

    //////////////////////
    // Save output mesh //
    //////////////////////

    lagrange::logger().info(
        "Output mesh has {} vertices and {} facets",
        mesh.get_num_vertices(),
        mesh.get_num_facets());

    lagrange::logger().info("Saving output mesh: {}", args.output);
    lagrange::io::save_mesh(args.output, mesh);

    return 0;
}

File: modules/subdivision/python/scripts/mesh_subdivision.py
#!/usr/bin/env python

#
# Copyright 2024 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.
#
"""Refine a mesh using subdivision surfaces."""

import argparse
import pathlib
import typing
import math
import logging

import numpy
import lagrange


def parse_args():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--num-levels", "-n", type=int, help="Maximum number of subdivision levels", default=4
    )
    parser.add_argument(
        "--edge-length",
        "-e",
        type=float,
        default=None,
        help="Maximum edge length to use with adaptive refinement (default: automatic based on max subdiv level).",
    )
    parser.add_argument(
        "--autosmooth-threshold",
        "-a",
        type=float,
        default=None,
        help="Normal angle threshold (in degree) for autodetecting sharp edges",
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument("--adaptive", action="store_true", dest="adaptive", default=True)
    group.add_argument("--uniform", action="store_false", dest="adaptive")
    parser.add_argument("input_mesh", type=pathlib.Path, help="Input mesh file")
    parser.add_argument(
        "output_mesh", type=pathlib.Path, default="output.obj", nargs="?", help="Output mesh file"
    )
    return parser.parse_args()


def prepare_asset(mesh: lagrange.SurfaceMesh, autosmooth: typing.Optional[float]):
    # Weld indexed attributes
    for id in mesh.get_matching_attribute_ids(lagrange.AttributeElement.Indexed):
        lagrange.weld_indexed_attribute(mesh, id, epsilon_rel=1e-3, epsilon_abs=1e-3)

    # Find an attribute to use as normals to define sharp edges
    normal_id = mesh.get_matching_attribute_id(usage=lagrange.AttributeUsage.Normal)
    if normal_id:
        lagrange.logger.info("Found normal attribute: %s", mesh.get_attribute_name(normal_id))

    # If no normals are found, compute autosmooth normals if requested
    if autosmooth is not None and normal_id:
        lagrange.logger.info(
            "Computing autosmooth normals with a threshold of %s degrees", autosmooth
        )
        normal_id = lagrange.compute_normal(mesh, math.radians(autosmooth))

    if normal_id:
        # Compute edge sharpness based on indexed normal topology
        seam_id = lagrange.compute_seam_edges(mesh, normal_id)
        edge_sharpness_id = lagrange.cast_attribute(mesh, seam_id, dtype=float)

        # Set vertex sharpness to 1 for leaf and junction vertices
        valence_id = lagrange.compute_vertex_valence(
            mesh, induced_by_attribute=mesh.get_attribute_name(seam_id)
        )
        vertex_sharpness_id = mesh.create_attribute(
            "vertex_sharpness",
            element=lagrange.AttributeElement.Vertex,
            num_channels=1,
            dtype=float,
        )
        val = mesh.attribute(valence_id).data
        mesh.attribute(vertex_sharpness_id).data[:] = numpy.select([val == 1, val > 2], [1, 1], 0)
    else:
        edge_sharpness_id = None
        vertex_sharpness_id = None

    return edge_sharpness_id, vertex_sharpness_id


def main():
    args = parse_args()
    lagrange.logger.setLevel(logging.INFO)

    # Load input mesh
    lagrange.logger.info("Loading input mesh: %s", args.input_mesh)
    if args.input_mesh.suffix in [".gltf", ".glb"]:
        lagrange.logger.warning(
            "Input mesh is a glTF file. Essential connectivity information is lost when loading from a glTF asset. We strongly advise using .fbx or .obj as input file format rather than glTF."
        )
    mesh = lagrange.io.load_mesh(args.input_mesh, stitch_vertices=True)
    lagrange.logger.info(
        "Input mesh has %s vertices and %s facets", mesh.num_vertices, mesh.num_facets
    )

    # Prepare sharpness attributes
    edge_sharpness_id, vertex_sharpness_id = prepare_asset(mesh, args.autosmooth_threshold)
    lagrange.logger.info(
        "Running subdivision with %s levels, target edge length: %s",
        args.num_levels,
        args.edge_length,
    )

    # Perform subdivision
    for id in mesh.get_matching_attribute_ids():
        if mesh.is_attribute_indexed(id):
            lagrange.logger.info(
                "Attribute %s has dtype=%s",
                mesh.get_attribute_name(id),
                mesh.indexed_attribute(id).values.dtype,
            )
        else:
            lagrange.logger.info(
                "Attribute %s has dtype=%s", mesh.get_attribute_name(id), mesh.attribute(id).dtype
            )
    mesh = lagrange.subdivision.subdivide_mesh(
        mesh,
        num_levels=args.num_levels,
        adaptive=args.adaptive,
        edge_sharpness_attr=edge_sharpness_id,
        vertex_sharpness_attr=vertex_sharpness_id,
        use_limit_surface=args.adaptive,
    )

    # Save result
    lagrange.logger.info(
        "Output mesh has %s vertices and %s facets", mesh.num_vertices, mesh.num_facets
    )
    lagrange.io.save_mesh(args.output_mesh, mesh)


if __name__ == "__main__":
    main()
